src/GraphQL.Tests/Types/ComplexGraphTypeTests.cs

```diff
@@ -468,4 +470,49 @@ public void should_not_throw_exception_on_valid_field_name_using_field_builder(s
 
         type.Fields.Last().Name.ShouldBe(fieldName);
     }
+
+    [Fact]
+    public void create_fieldbuilder_with_inferred_graph_type()
+    {
+        var type = new ComplexType<TestObject>();
+
+        type.Field<int>("field1").Resolve(_ => 3);
+        type.Fields.Find("field1").ShouldNotBeNull().Type.ShouldBe(typeof(NonNullGraphType<GraphQLClrOutputTypeReference<int>>));
+
+        type.Field<int>("field2", false).Resolve(_ => 3);
+        type.Fields.Find("field2").ShouldNotBeNull().Type.ShouldBe(typeof(NonNullGraphType<GraphQLClrOutputTypeReference<int>>));
+
+        type.Field<int>("field3", true).Resolve(_ => 3);
+        type.Fields.Find("field3").ShouldNotBeNull().Type.ShouldBe(typeof(GraphQLClrOutputTypeReference<int>));
+
+        Should.Throw<ArgumentException>(() => type.Field<int?>("field4"));
+        Should.Throw<ArgumentException>(() => type.Field<int?>("field5", false));
+
+        type.Field<int?>("field6", true).Resolve(_ => 3);
+        type.Fields.Find("field6").ShouldNotBeNull().Type.ShouldBe(typeof(GraphQLClrOutputTypeReference<int>));
+
+        // note: NRT attributes cannot be inferred/read below as they are not actually stored in the compiled code
+
+        type.Field<string>("field7").Resolve(_ => "hello");
+        type.Fields.Find("field7").ShouldNotBeNull().Type.ShouldBe(typeof(NonNullGraphType<GraphQLClrOutputTypeReference<string>>));
+
+        type.Field<string>("field8", false).Resolve(_ => "hello");
+        type.Fields.Find("field8").ShouldNotBeNull().Type.ShouldBe(typeof(NonNullGraphType<GraphQLClrOutputTypeReference<string>>));
+
+        type.Field<string>("field9", true).Resolve(_ => "hello");
+        type.Fields.Find("field9").ShouldNotBeNull().Type.ShouldBe(typeof(GraphQLClrOutputTypeReference<string>));
+
+        type.Field<string?>("field10").Resolve(_ => "hello");
+        type.Fields.Find("field10").ShouldNotBeNull().Type.ShouldBe(typeof(NonNullGraphType<GraphQLClrOutputTypeReference<string>>));
+
+        type.Field<string?>("field11", false).Resolve(_ => "hello");
+        type.Fields.Find("field11").ShouldNotBeNull().Type.ShouldBe(typeof(NonNullGraphType<GraphQLClrOutputTypeReference<string>>));
+
+        type.Field<string?>("field12", true).Resolve(_ => "hello");
+        type.Fields.Find("field12").ShouldNotBeNull().Type.ShouldBe(typeof(GraphQLClrOutputTypeReference<string>));
+
+        // notice here that since the NRT attribute of 'string' cannot be read, it is assumed to be nullable
```

Comment:
> notice here that since the NRT attribute of 'string' cannot be read

Why?

Reply:
NRT annotations are stored on method signatures (a `MethodInfo` / similar) or generic class definitions.  An inline function doesn't have a `MethodInfo` to read, I believe, and a call to a generic function is directly compiled; no NRT attributes are stored anywhere.

So in the following example, NRT annotations are dropped by the compiler and there is no way to recover them:

```cs
Field<string?>("test");
```

I could be wrong, but that's my understanding.

Reply:
Because of the same issue, some class definitions cannot properly determine NRT status:

```cs
class ClassA<T>
{
    public T GetValue();
}

class ClassB : ClassA<string?> // these NRT attributes are stored
{
}

var valueA = new ClassA<string?>().GetValue();  // GetValue cannot determine NRT status of T
var valueB = new ClassB().GetValue();  // GetValue could determine NRT status of T
```

Reply:
![изображение](https://user-images.githubusercontent.com/21261007/180449575-3b9a75b2-2455-4ff8-963a-9ce2b4c5b7d0.png)
😕

Reply:
Of course the compiler knows the type.  It's just not saved into the dll, so reflection cannot see it.

Reply:
ok